### PR TITLE
Add logs to script and command execution

### PR DESCRIPTION
**Pull request by stephankruggg**
### Description

This PR adds and improves existing logs when executing scripts and commands, since current information is not sufficient to enable troubleshooting. 

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [X] Minor

### How Has This Been Tested?

I applied the changes in a local lab and checked if the updated logs were successfully logged.

### CODE REVIEW

**Review by BryanMLima — APPROVED** [review]
LGTM

File: utils/src/main/java/com/cloud/utils/script/Script.java [hunk 1]
```diff
@@ -205,19 +205,20 @@ static String stackTraceAsString(Throwable throwable) {
     public String execute(OutputInterpreter interpreter) {
         String[] command = _command.toArray(new String[_command.size()]);
 
-        if (_logger.isDebugEnabled()) {
-            _logger.debug("Executing: " + buildCommandLine(command).split(KeyStoreUtils.KS_FILENAME)[0]);
-        }
+        String commandLine = buildCommandLine(command);
+        _logger.debug(String.format("Executing command [%s].", commandLine.split(KeyStoreUtils.KS_FILENAME)[0]));
```
Comment (DaanHoogland) [inline utils/src/main/java/com/cloud/utils/script/Script.java:209]:
```suggestion
        String commandLine = buildCommandLine(command);
        if (_logger.isDebugEnabled()) {
            _logger.debug(String.format("Executing command [%s].", commandLine.split(KeyStoreUtils.KS_FILENAME)[0]));
        }
```

File: utils/src/main/java/com/cloud/utils/script/Script.java [hunk 2]
```diff
@@ -205,19 +205,20 @@ static String stackTraceAsString(Throwable throwable) {
...
 
         try {
+            _logger.trace(String.format("Creating process for command [%s].", commandLine));
```
Comment (DaanHoogland) [inline utils/src/main/java/com/cloud/utils/script/Script.java:212]:
```suggestion
            if (_logger.isTraceEnabled()) {
                _logger.trace(String.format("Creating process for command [%s].", commandLine));
            }
```

File: utils/src/main/java/com/cloud/utils/script/Script.java [hunk 3]
```diff
@@ -205,19 +205,20 @@ static String stackTraceAsString(Throwable throwable) {
...
             ProcessBuilder pb = new ProcessBuilder(command);
             pb.redirectErrorStream(true);
             if (_workDir != null)
                 pb.directory(new File(_workDir));
 
+            _logger.trace(String.format("Starting process for command [%s].", commandLine));
```
Comment (DaanHoogland) [inline utils/src/main/java/com/cloud/utils/script/Script.java:218]:
```suggestion
            if (_logger.isTraceEnabled()) {
                _logger.trace(String.format("Starting process for command [%s].", commandLine));
            }
```

File: utils/src/main/java/com/cloud/utils/script/Script.java [hunk 4]
```diff
@@ -226,40 +227,48 @@ public String execute(OutputInterpreter interpreter) {
             _thread = Thread.currentThread();
             ScheduledFuture<String> future = null;
             if (_timeout > 0) {
+                _logger.trace(String.format("Scheduling the execution of command [%s] with a timeout of [%s] milliseconds.", commandLine, _timeout));
```
Comment (DaanHoogland) [inline utils/src/main/java/com/cloud/utils/script/Script.java:230]:
```suggestion
                if (_logger.isTraceEnabled() ) {
                    _logger.trace(String.format("Scheduling the execution of command [%s] with a timeout of [%s] milliseconds.", commandLine, _timeout));
                }
```

File: utils/src/main/java/com/cloud/utils/script/Script.java [hunk 5]
```diff
@@ -226,40 +227,48 @@ public String execute(OutputInterpreter interpreter) {
...
                 future = s_executors.schedule(this, _timeout, TimeUnit.MILLISECONDS);
             }
 
+            long processPid = _process.pid();
             Task task = null;
             if (interpreter != null && interpreter.drain()) {
+                _logger.trace(String.format("Executing interpreting task of process [%s] for command [%s].", processPid, commandLine));
```
Comment (DaanHoogland) [inline utils/src/main/java/com/cloud/utils/script/Script.java:237]:
```suggestion
                if (_logger.isTraceEnabled()) {
                    _logger.trace(String.format("Executing interpreting task of process [%s] for command [%s].", processPid, commandLine));
                }
```

File: utils/src/main/java/com/cloud/utils/script/Script.java [hunk 6]
```diff
@@ -226,40 +227,48 @@ public String execute(OutputInterpreter interpreter) {
...
                 task = new Task(interpreter, ir);
                 s_executors.execute(task);
             }
 
             while (true) {
-                _logger.debug("Executing while with timeout : " + _timeout);
+                _logger.trace(String.format("Attempting process [%s] execution for command [%s] with timeout [%s].", processPid, commandLine, _timeout));
```
Comment (DaanHoogland) [inline utils/src/main/java/com/cloud/utils/script/Script.java:243]:
```suggestion
                if (_logger.isTraceEnabled()) {
                    _logger.trace(String.format("Attempting process [%s] execution for command [%s] with timeout [%s].", processPid, commandLine, _timeout));
                }
```

File: utils/src/main/java/com/cloud/utils/script/Script.java [hunk 7]
```diff
@@ -226,40 +227,48 @@ public String execute(OutputInterpreter interpreter) {
...
                 try {
-                    //process execution completed within timeout period
                     if (_process.waitFor(_timeout, TimeUnit.MILLISECONDS)) {
-                        //process completed successfully
+                        _logger.trace(String.format("Process [%s] execution for command [%s] completed within timeout period [%s].", processPid, commandLine,
+                                _timeout));
```
Comment (DaanHoogland) [inline utils/src/main/java/com/cloud/utils/script/Script.java:247]:
```suggestion
                        if (_logger.isTraceEnabled()) {
                            _logger.trace(String.format("Process [%s] execution for command [%s] completed within timeout period [%s].", processPid, commandLine,
                                    _timeout));
                        }
```

File: utils/src/main/java/com/cloud/utils/script/Script.java [hunk 8]
```diff
@@ -226,40 +227,48 @@ public String execute(OutputInterpreter interpreter) {
...
                         if (_process.exitValue() == 0) {
-                            _logger.debug("Execution is successful.");
+                            _logger.debug(String.format("Successfully executed process [%s] for command [%s].", processPid, commandLine));
                             if (interpreter != null) {
-                                return interpreter.drain() ? task.getResult() : interpreter.interpret(ir);
+                                if (interpreter.drain()) {
+                                    _logger.trace(String.format("Returning task result of process [%s] for command [%s].", processPid, commandLine));
```
Comment (DaanHoogland) [inline utils/src/main/java/com/cloud/utils/script/Script.java:252]:
```suggestion
                                    if (_logger.isTraceEnabled()) {
                                        _logger.trace(String.format("Returning task result of process [%s] for command [%s].", processPid, commandLine));
                                    }
```

File: utils/src/main/java/com/cloud/utils/script/Script.java [hunk 9]
```diff
@@ -226,40 +227,48 @@ public String execute(OutputInterpreter interpreter) {
...
+                                    return task.getResult();
+                                }
+                                _logger.trace(String.format("Returning interpretation of process [%s] for command [%s].", processPid, commandLine));
```
Comment (DaanHoogland) [inline utils/src/main/java/com/cloud/utils/script/Script.java:255]:
```suggestion
                                if (_logger.isTraceEnabled()) {
                                    _logger.trace(String.format("Returning interpretation of process [%s] for command [%s].", processPid, commandLine));
                                }
```

**Comment by stephankruggg** [comment]
@DaanHoogland, as discussed [here](https://github.com/apache/cloudstack/pull/4707#discussion_r578495829), the isEnabled check is detrimental in terms of code legibility. In my opinion, this check's barely noticeable performance gain is not sufficient to counter this negative impact, so it should not be used in this scenario.

**Comment by DaanHoogland** [comment]
> @DaanHoogland, the isEnabled check is detrimental in terms of code legibility. In my opinion, this check's barely noticeable performance gain is not sufficient to counter this negative impact, so it should not be used in this scenario.

I don´t agree. we could change the log library to add variable parameters to the trace/debug/info messages to make the log line look cleaner. I´m very extrmely for legibility but not at cost of neat following of best practices.

**Comment by stephankruggg** [comment]
> > @DaanHoogland, the isEnabled check is detrimental in terms of code legibility. In my opinion, this check's barely noticeable performance gain is not sufficient to counter this negative impact, so it should not be used in this scenario.
> 
> I don´t agree. we could change the log library to add variable parameters to the trace/debug/info messages to make the log line look cleaner. I´m very extrmely for legibility but not at cost of neat following of best practices.

I understand; however, it is not consensus that the isEnabled check should be used, and the scenarios here do not require so much processing in order to make it necessary to validate the log level before creating the string, so I rather not add them.

**Review by JoaoJandre — APPROVED** [review]
CLGTM

**Comment by GaOrtiga** [comment]
As mentioned in comment https://github.com/apache/cloudstack/pull/6947#issuecomment-1611851403, I will be helping out @stephankruggg with some of his opened PRs.

@DaanHoogland given that the patch for [upgrading Log4j ](https://github.com/apache/cloudstack/pull/7131) is already ready, is guarding these logs still necessary? Could we move on with this PR?

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Review by shwstppr — APPROVED** [review]
Code LGTM

**Comment by DaanHoogland** [comment]
I am not +1, more like -0, but too many do not agree with my stance on the logging issue, so I'll merge this anyway.

**Pull request merged**